src/compute/src/sink/persist_sink.rs

```diff
@@ -570,258 +568,313 @@ where
+        let mut correction = Vec::new();
+
+        // Contains descriptions of batches for which we know that we can
+        // write data. We got these from the "centralized" operator that
+        // determines batch descriptions for all writers.
+        let mut in_flight_batches: HashMap<
+            (Antichain<Timestamp>, Antichain<Timestamp>),
+            Capability<Timestamp>,
+        > = HashMap::new();
+
+        // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+        let persist_client = persist_clients
+            .lock()
+            .await
+            .open(persist_location)
+            .await
+            .expect("could not open persist client");
+
+        let mut write = persist_client
+            .open_writer::<SourceData, (), Timestamp, Diff>(
+                shard_id,
+                &format!("persist_sink::write_batches {}", sink_id),
+            )
+            .await
+            .expect("could not open persist shard");
+
+        // The current input frontiers.
+        let mut batch_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut persist_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // Keep track of alive inputs, so we know when we can shut down the operator.
+        let mut descriptions_input_alive = true;
+        let mut desired_input_alive = true;
+        let mut persist_input_alive = true;
+
+        macro_rules! shutdown_on_dead_inputs {
+            () => {
+                if !descriptions_input_alive && !desired_input_alive && !persist_input_alive {
                     return;
                 }
+            };
+        }
+
+        let mut ticker = tokio::time::interval(Duration::from_secs(60));
+        ticker.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+        tokio::pin!(shutdown_rx);
+        loop {
+            // Make sure that our write handle lease does not expire!
+            //
+            // We don't write to _Consensus_ but instead only write batch
+            // data to _Blob_ so someone might eventually expire our lease.
+            // This could lead to the "leaked blob reaper" eventually
+            // removing blobs that have been written but not yet appended to
+            // the shard.
+            //
+            write.maybe_heartbeat_writer().await;
+            trace!("persist_sink {sink_id}/{shard_id}: writer heartbeating write handle");
+
+            tokio::select! {
+                // Always check for shutdown first.
+                biased;
+                _ = shutdown_rx.as_mut() => return,
+
+                // Ensure that we are periodically woken up and keep heartbeating the persist
+                // writer even if there are no changes in our inputs (updates or frontier changes).
+                _ = ticker.tick() => continue,
```

Comment:
Could move the `write.maybe_heartbeat_writer().await;` to here.

Reply:
I was not sure about this. The comment above states that "We don't write to _Consensus_ but instead only write batch data to _Blob_ so someone might eventually expire our lease." I take that to mean that even if we see progress on the inputs, we still should heartbeat the writer? If that is the case, then we can't move the heartbeating here, because this branch is only taken if nothing else happens for longer than 60 seconds, which might never be the case for sufficiently busy inputs.

Reply:
Doesn't matter now that I have removed the heartbeating anyway!